### DISCUSSION TITLE
Wrong role claim on PostgREST

### DISCUSSION THREAD

**Discussion by andreivreja** [supabase/supabase#614] [Questions] [answered]
Hi there,

First of all, congrats on the amazing work you guys have done on Supabase, it’s pretty straightforward!

Having an issue with PostgREST auth (managed Supabase), I’m using the default service_role token to insert data (I wanna lock anon to read only), however, somehow it’s being detected as anon.

At least that’s the conclusion I came to after this behavior:
1. Enabling RLS results in permission error with service_role token (which should bypass RLS).
2. Giving access to role anon on the policy makes it so the insert works, which I guess confirms that somehow it’s using (or defaulting?) anon role.

I can confirm that I’m using the correct token. Not using one would result an authorization failure (from the PostgREST docs, it should default to the anon role; unless that’s an exception in production).

Can’t really seem track this issue down on my own.

Note: haven’t tried generating a new JWT with service_role as the role claim, only tried the default one from the dashboard, but since it’s a JWT this shouldn’t be relevant.

**Comment by steve-chavez** [comment, marked as answer]
Hey @andreivreja,

Can you double check you're using the service_role JWT? You can paste it in [jwt.io](https://jwt.io/) and see if the `role` claim has `service_role`.

Also note that you shouldn't be using `service_role` for clients.

**Reply by andreivreja** [reply]
It’s indeed the service_role JWT. I plan to use it on the backend for updating the DB, while using the anon one with read only access for a realtime stream on the client.

**Reply by steve-chavez** [reply]
Can you check the service role has bypassrls? Try this query:

```sql
select rolname, rolbypassrls from pg_roles where rolname = 'service_role';
```

`rolbypassrls` should show `true`.

**Reply by andreivreja** [reply]
It is indeed true. As mentioned in the first post, if my RLS policy checks if the role is `anon`, that evaluates to true so for sure somehow the role assignment is failing. I made sure `authenticator` has permission to switch to `service_role`.

I’ll set up a new project in a little bit to rule out the possibility of my changes potentially messing it up; however, I started changing stuff only after I found out the issue with role claim.

![B430DF09-AA81-41EF-BC7D-E9DD5E240811](https://user-images.githubusercontent.com/8231635/106034077-1e74af80-60db-11eb-891f-511a81dfef2e.jpeg)

**Reply by steve-chavez** [reply]
> that evaluates to true so for sure somehow the role assignment is failing

PostgREST would show an error if that's the case. To double check that this isn't happening:

```sql
create or replace function showrole() returns text as $$
  select current_setting('role', true);
$$ language sql;
```

Then call it through `/rpc/showrole`. It should show `service_role` if that's indeed in the JWT role claim.

**Reply by andreivreja** [reply]
Shows `anon` for both `anon` and `service_role`. Another weird thing is, I just tried generating a new JWT, I literally just added 1 to the `exp` claim on the default `service_role` token, signed it with the same secret, and I get this

```
{
  error: { message: 'Invalid authentication credentials' },
  data: null,
  count: null,
  status: 401,
  statusText: 'Unauthorized',
  body: null
}
```

The payload for the token is
```
{
  "role": "service_role",
  "iat": 1611523691,
  "exp": 1927099692
}
```

Apparently, rather than verifying and decoding the token, it literally checks for an exact string match on the token.

**Reply by andreivreja** [reply]
Just created a fresh project, added your `showrole` example function, used `service_role` token and it's still being seen as `anon`.

**Reply by steve-chavez** [reply]
> Just created a fresh project, added your showrole example function, used service_role token and it's still being seen as anon.

Hmm.. I can't reproduce. Where are you getting the JWT? 

I get `curl` calls samples with the different role JWTs by clicking on the API tab here:

![2021-01-27-144751_538x47_scrot](https://user-images.githubusercontent.com/1829294/106045238-b3ad7f00-60ae-11eb-8c8d-01a2a3ae4487.png)

That's on `{your_project}/api/default?rpc=showrole`.

**Reply by andreivreja** [reply]
Oh, so I must include the token both as an `apikey` and as a bearer token for `authorization`? Didn't even cross my mind until you directed me to the auto generated snippets on the API tab.

**Reply by andreivreja** [reply]
Okay yeah, everything works as it's supposed to now. I'm not using the `supabase-js`, had errors saying that I didn't include an `apikey` after trying with an auth header, so I assumed it's just a different name, turns out I need to use both headers. Is this some inconsistency or it was just more straightforward to have it this way to account for all uses cases?

Either way, thanks for pointing out the API code snippets, would've spent a lot of time to figure this one out otherwise since the docs/repos didn't really mention anything about this case; or maybe I just didn't find the info.

Keep up the amazing work!

**Reply by kiwicopple** [reply]
> Is this some inconsistency or it was just more straightforward to have it this way to account for all uses cases?

Yeah we had to do this because when we first lauched we didnt have Supabase Auth - only an API key for the API gateway. We added GoTrue + RLS (via PostgREST) for Supabase Auth, and so the "anon" JWT also became API key. We will probably remove this at some stage for simplicity, but for now the API key is useful because it prevents the database from getting DOS'd by unauthenticated requests.

> Keep up the amazing work!

Thanks for your kind words @andreivreja

**Reply by mstade** [reply]
Thanks for that explanation @kiwicopple, I've also been wondering why the `apikey` header exists. Seems to me that at this point though you could just get rid of it and rely only on JWTs, no?

**Reply by kiwicopple** [reply]
Very soon! We'll phase it out incrementally from our libraries + backend
